File: integrations/overmind/handlers.go
package overmind

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
	"strconv"

	mcp "github.com/daltoniam/switchboard"
)

func listAvailableAgents(ctx context.Context, o *overmind, args map[string]any) (*mcp.ToolResult, error) {
	data, err := o.get(ctx, "/api/flow_runs/%s/available_agents", url.PathEscape(o.flowRunID))
	if err != nil {
		return mcp.ErrResult(err)
	}
	return mcp.RawResult(data)
}

func launchAgent(ctx context.Context, o *overmind, args map[string]any) (*mcp.ToolResult, error) {
	agentID := argStr(args, "agent_id")
	if agentID == "" {
		return mcp.ErrResult(fmt.Errorf("agent_id is required"))
	}
	prompt := argStr(args, "prompt")
	if prompt == "" {
		return mcp.ErrResult(fmt.Errorf("prompt is required"))
	}
	agentContext := argStr(args, "context")

	body := map[string]any{
		"agent_id":      agentID,
		"prompt":        prompt,
		"parent_run_id": o.agentRunID,
	}
	if agentContext != "" {
		body["context"] = agentContext
	}

	data, err := o.post(ctx, fmt.Sprintf("/api/flow_runs/%s/launch_agent", url.PathEscape(o.flowRunID)), body)
	if err != nil {
		return mcp.ErrResult(err)
	}
	return mcp.RawResult(data)
}

func getAgentStatus(ctx context.Context, o *overmind, args map[string]any) (*mcp.ToolResult, error) {
	agentRunID := argStr(args, "agent_run_id")
	if agentRunID == "" {
		return mcp.ErrResult(fmt.Errorf("agent_run_id is required"))
	}

	data, err := o.get(ctx, "/api/agent_runs/%s/status", url.PathEscape(agentRunID))
	if err != nil {
		return mcp.ErrResult(err)
	}
	return mcp.RawResult(data)
}

func getAgentResult(ctx context.Context, o *overmind, args map[string]any) (*mcp.ToolResult, error) {
	agentRunID := argStr(args, "agent_run_id")
	if agentRunID == "" {
		return mcp.ErrResult(fmt.Errorf("agent_run_id is required"))
	}

	data, err := o.get(ctx, "/api/agent_runs/%s/result", url.PathEscape(agentRunID))
	if err != nil {
		return mcp.ErrResult(err)
	}
	return mcp.RawResult(data)
}

func completeFlow(ctx context.Context, o *overmind, args map[string]any) (*mcp.ToolResult, error) {
	summary := argStr(args, "summary")
	if summary == "" {
		return mcp.ErrResult(fmt.Errorf("summary is required"))
	}
	status := argStr(args, "status")
	if status == "" {
		status = "success"
	}
	if status != "success" && status != "failure" {
		return mcp.ErrResult(fmt.Errorf("status must be 'success' or 'failure', got %q", status))
	}

	body := map[string]any{
		"summary":      summary,
		"status":       status,
		"agent_run_id": o.agentRunID,
	}

	data, err := o.post(ctx, fmt.Sprintf("/api/flow_runs/%s/complete", url.PathEscape(o.flowRunID)), body)
	if err != nil {
		return mcp.ErrResult(err)
	}
	return mcp.RawResult(data)
}

// --- CRUD handlers ---

func listMCPRoles(ctx context.Context, o *overmind, _ map[string]any) (*mcp.ToolResult, error) {
	data, err := o.get(ctx, "/api/mcp_roles")
	if err != nil {
		return mcp.ErrResult(err)
	}
	return mcp.RawResult(data)
}

func createMCPRole(ctx context.Context, o *overmind, args map[string]any) (*mcp.ToolResult, error) {
	name := argStr(args, "name")
	if name == "" {
		return mcp.ErrResult(fmt.Errorf("name is required"))
	}

	body := map[string]any{"name": name}
	if desc := argStr(args, "description"); desc != "" {
		body["description"] = desc
	}

	data, err := o.post(ctx, "/api/mcp_roles", body)
	if err != nil {
		return mcp.ErrResult(err)
	}
	return mcp.RawResult(data)
}

func listAgents(ctx context.Context, o *overmind, _ map[string]any) (*mcp.ToolResult, error) {
	data, err := o.get(ctx, "/api/agents")
	if err != nil {
		return mcp.ErrResult(err)
	}
	return mcp.RawResult(data)
}

func createAgent(ctx context.Context, o *overmind, args map[string]any) (*mcp.ToolResult, error) {
	name := argStr(args, "name")
	if name == "" {
		return mcp.ErrResult(fmt.Errorf("name is required"))
	}
	basePrompt := argStr(args, "base_prompt")
	if basePrompt == "" {
		return mcp.ErrResult(fmt.Errorf("base_prompt is required"))
	}
	model := argStr(args, "model")
	if model == "" {
		return mcp.ErrResult(fmt.Errorf("model is required"))
	}
	provider := argStr(args, "model_provider")
	if provider == "" {
		return mcp.ErrResult(fmt.Errorf("model_provider is required"))
	}
	roleID := argStr(args, "mcp_role_id")
	if roleID == "" {
		return mcp.ErrResult(fmt.Errorf("mcp_role_id is required"))
	}

	body := map[string]any{
		"name":           name,
		"base_prompt":    basePrompt,
		"model":          model,
		"model_provider": provider,
		"mcp_role_id":    roleID,
	}
	if desc := argStr(args, "description"); desc != "" {
		body["description"] = desc
	}

	data, err := o.post(ctx, "/api/agents", body)
	if err != nil {
		return mcp.ErrResult(err)
	}
	return mcp.RawResult(data)
}

func listFlows(ctx context.Context, o *overmind, _ map[string]any) (*mcp.ToolResult, error) {
	data, err := o.get(ctx, "/api/flows")
	if err != nil {
		return mcp.ErrResult(err)
	}
	return mcp.RawResult(data)
}

func createFlow(ctx context.Context, o *overmind, args map[string]any) (*mcp.ToolResult, error) {
	name := argStr(args, "name")
	if name == "" {
		return mcp.ErrResult(fmt.Errorf("name is required"))
	}
	agentID := argStr(args, "initial_agent_id")
	if agentID == "" {
		return mcp.ErrResult(fmt.Errorf("initial_agent_id is required"))
	}
	promptTemplate := argStr(args, "prompt_template")
	if promptTemplate == "" {
		return mcp.ErrResult(fmt.Errorf("prompt_template is required"))
	}

	body := map[string]any{
		"name":             name,
		"initial_agent_id": agentID,
		"prompt_template":  promptTemplate,
		"enabled":          true,
	}
	if desc := argStr(args, "description"); desc != "" {
		body["description"] = desc
	}

	if wp := argStr(args, "webhook_params"); wp != "" {
		var parsed json.RawMessage
		if json.Unmarshal([]byte(wp), &parsed) == nil {
			body["webhook_params"] = parsed
		}
	}
	if op := argStr(args, "output_params"); op != "" {
		var parsed json.RawMessage
		if json.Unmarshal([]byte(op), &parsed) == nil {
			body["output_params"] = parsed
		}
	}
	if e := argStr(args, "enabled"); e == "false" {
		body["enabled"] = false
	}
	if tm := argStr(args, "timeout_minutes"); tm != "" {
		if v, err := strconv.Atoi(tm); err == nil && v > 0 {
			body["timeout_minutes"] = v
		}
	}

	data, err := o.post(ctx, "/api/flows", body)
	if err != nil {
		return mcp.ErrResult(err)
	}
	return mcp.RawResult(data)
}

File: wasm/guest-rust/overmind/src/handlers.rs
use std::collections::HashMap;
use switchboard_guest_sdk as sdk;

use crate::{agent_run_id_str, do_get, do_post, flow_run_id_str, url_encode};

pub fn list_available_agents(_args: HashMap<String, serde_json::Value>) -> sdk::ToolResult {
    match do_get(&format!("/api/flow_runs/{}/available_agents", url_encode(&flow_run_id_str()))) {
        Ok(data) => sdk::raw_result(data),
        Err(e) => sdk::err_result(&e),
    }
}

pub fn launch_agent(args: HashMap<String, serde_json::Value>) -> sdk::ToolResult {
    let agent_id = sdk::arg_str(&args, "agent_id");
    let prompt = sdk::arg_str(&args, "prompt");
    let context = sdk::arg_str(&args, "context");
    if agent_id.is_empty() {
        return sdk::err_result("agent_id is required");
    }
    if prompt.is_empty() {
        return sdk::err_result("prompt is required");
    }
    let mut body = serde_json::json!({
        "agent_id": agent_id,
        "prompt": prompt,
        "parent_run_id": agent_run_id_str(),
    });
    if !context.is_empty() {
        body["context"] = serde_json::Value::String(context);
    }
    match do_post(&format!("/api/flow_runs/{}/launch_agent", url_encode(&flow_run_id_str())), Some(body)) {
        Ok(data) => sdk::raw_result(data),
        Err(e) => sdk::err_result(&e),
    }
}

pub fn get_agent_status(args: HashMap<String, serde_json::Value>) -> sdk::ToolResult {
    let id = sdk::arg_str(&args, "agent_run_id");
    if id.is_empty() {
        return sdk::err_result("agent_run_id is required");
    }
    match do_get(&format!("/api/agent_runs/{}/status", url_encode(&id))) {
        Ok(data) => sdk::raw_result(data),
        Err(e) => sdk::err_result(&e),
    }
}

pub fn get_agent_result(args: HashMap<String, serde_json::Value>) -> sdk::ToolResult {
    let id = sdk::arg_str(&args, "agent_run_id");
    if id.is_empty() {
        return sdk::err_result("agent_run_id is required");
    }
    match do_get(&format!("/api/agent_runs/{}/result", url_encode(&id))) {
        Ok(data) => sdk::raw_result(data),
        Err(e) => sdk::err_result(&e),
    }
}

pub fn complete_flow(args: HashMap<String, serde_json::Value>) -> sdk::ToolResult {
    let summary = sdk::arg_str(&args, "summary");
    let mut status = sdk::arg_str(&args, "status");
    if summary.is_empty() {
        return sdk::err_result("summary is required");
    }
    if status.is_empty() {
        status = "success".into();
    }
    if status != "success" && status != "failure" {
        return sdk::err_result(&format!("status must be 'success' or 'failure', got {:?}", status));
    }
    let body = serde_json::json!({
        "summary": summary,
        "status": status,
        "agent_run_id": agent_run_id_str(),
    });
    match do_post(&format!("/api/flow_runs/{}/complete", url_encode(&flow_run_id_str())), Some(body)) {
        Ok(data) => sdk::raw_result(data),
        Err(e) => sdk::err_result(&e),
    }
}
